From 5662452a219adc21d7515014439e663e0475d17c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Cl=C3=A9ment=20Le=20Goffic?= <clement.legoffic@foss.st.com>
Date: Wed, 19 Feb 2025 15:25:18 +0100
Subject: [PATCH] perf: stm32: set RAM type when resuming from low power
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The DRAMINF register value is not saved when going in low power.
When resuming this register must be considered cleared.
Hence, implement a resume callback to restore DRAMINF value.

Change-Id: I66171a025e14780ece41ad54ac79ad4a174a7c2a
Signed-off-by: Clément Le Goffic <clement.legoffic@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/438464
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Antonio Maria BORNEO <antonio.borneo@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Antonio Maria BORNEO <antonio.borneo@foss.st.com>
---
 drivers/perf/stm32_ddr_pmu.c | 23 +++++++++++++++++++----
 1 file changed, 19 insertions(+), 4 deletions(-)

diff --git a/drivers/perf/stm32_ddr_pmu.c b/drivers/perf/stm32_ddr_pmu.c
index 1b7666c0edd2..75418120a98b 100644
--- a/drivers/perf/stm32_ddr_pmu.c
+++ b/drivers/perf/stm32_ddr_pmu.c
@@ -105,6 +105,7 @@ struct stm32_ddr_pmu {
 	struct hrtimer hrtimer;
 	ktime_t poll_period;
 	int selected_set;
+	u32 dram_type;
 	struct list_head counters[];
 };
 
@@ -758,9 +759,7 @@ static int stm32_ddr_pmu_device_probe(struct platform_device *pdev)
 	}
 
 	if (pmu->cfg->regs->dram_inf.reg) {
-		u32 dram_type;
-
-		ret = of_property_read_u32(pdev->dev.of_node, "st,dram-type", &dram_type);
+		ret = of_property_read_u32(pdev->dev.of_node, "st,dram-type", &pmu->dram_type);
 		if (ret) {
 			dev_err(&pdev->dev, "Missing device-tree property 'st,dram-type'\n");
 			perf_pmu_unregister(&pmu->pmu);
@@ -768,7 +767,7 @@ static int stm32_ddr_pmu_device_probe(struct platform_device *pdev)
 			return ret;
 		}
 
-		writel_relaxed(dram_type, pmu->membase + pmu->cfg->regs->dram_inf.reg);
+		writel_relaxed(pmu->dram_type, pmu->membase + pmu->cfg->regs->dram_inf.reg);
 	}
 
 	clk_disable(pmu->clk);
@@ -788,6 +787,17 @@ static void stm32_ddr_pmu_device_remove(struct platform_device *pdev)
 	perf_pmu_unregister(&stm32_ddr_pmu->pmu);
 }
 
+static int stm32_ddr_pmu_device_resume(struct device *dev)
+{
+	struct stm32_ddr_pmu *pmu = dev_get_drvdata(dev);
+
+	clk_enable(pmu->clk);
+	writel_relaxed(pmu->dram_type, pmu->membase + pmu->cfg->regs->dram_inf.reg);
+	clk_disable(pmu->clk);
+
+	return 0;
+}
+
 static const struct stm32_ddr_pmu_regspec stm32mp1_ddr_pmu_regspec = {
 	.stop =		{ DDRPERFM_CTRL, CTRL_STOP },
 	.start =	{ DDRPERFM_CTRL, CTRL_START },
@@ -846,6 +856,10 @@ static const struct stm32_ddr_pmu_cfg stm32mp2_ddr_pmu_cfg = {
 	.get_counter = stm32_ddr_pmu_get_event_counter_mp2,
 };
 
+static const struct dev_pm_ops stm32_ddr_pmu_pm_ops = {
+	SET_SYSTEM_SLEEP_PM_OPS(NULL, stm32_ddr_pmu_device_resume)
+};
+
 static const struct of_device_id stm32_ddr_pmu_of_match[] = {
 	{ .compatible = "st,stm32-ddr-pmu", .data = &stm32mp_ddr_pmu_cfg },
 	{ .compatible = "st,stm32mp25-ddr-pmu", .data = &stm32mp2_ddr_pmu_cfg },
@@ -856,6 +870,7 @@ MODULE_DEVICE_TABLE(of, stm32_ddr_pmu_of_match);
 static struct platform_driver stm32_ddr_pmu_driver = {
 	.driver = {
 		.name	= "stm32-ddr-pmu",
+		.pm = &stm32_ddr_pmu_pm_ops,
 		.of_match_table = of_match_ptr(stm32_ddr_pmu_of_match),
 	},
 	.probe = stm32_ddr_pmu_device_probe,
-- 
2.50.1

